### DISCUSSION TITLE
Optional inputs in v5

### DISCUSSION THREAD

**Discussion by AnatoleLucet** [statelyai/xstate#4819] [Q&A]
I'm trying to create a state machines which takes optional inputs, but I cannot figure out how to do this using the `input()` function.

```typescript
const test = setup({
  types: {} as {
    input?: {
      foo?: string;
    };
  },
}).createMachine({
  context: ({ input = {} }) => ({
    foo: input.foo ?? "bar",
  }),
});

createActor(test); // error missing "input" property
createActor(test, { input: undefined }); // no error but not very practical
```

**Comment by Andarist** [comment]
This one works with `exactOptionalPropertyTypes` (not that I recommend you enable this setting right now):
```ts
// @exactOptionalPropertyTypes: true
import { setup, createActor } from "xstate"; // types: 5.9.1

const test = setup({
  types: {} as {
    input:
      | {
          foo?: string;
        }
      | undefined;
  },
}).createMachine({
  context: ({ input = {} }) => ({
    foo: input.foo ?? "bar",
  }),
});

createActor(test);
createActor(test, { input: undefined });
createActor(test, { input: { foo: "" } });
```

[TS playground](https://www.typescriptlang.org/play?exactOptionalPropertyTypes=true#code/PTAEAEFMA8EMGMAuB5ADoglgewHawDYAKATlqpMYgJ4AqV5AzgFyiLECukAsAFAYC2qLJVABvUA0iJ2qADSh4xSLESQAgkmGgAvqABmpfqABE0BohWRjAblAhW9SM1ABWAHQBONwEZeveLjmrE6IoAC8ElIyABSivKAOjCyiurAMYvEJoBg4qOyITJlZoAA+GTzFlfpYWAD8LObEOQDm1kWV2u0JZew4ACaQejmQfW0VOrK82gCUborKqgCyCAAWw7GZATiq0AWgsdm5+eFiujPhAHz7ceMJejUsOXmIbvdYoLW1JgBGsMTGk3GM0BMzG-iUlg0iGE0VU5mmY3mkM0xFhIXk4ie+RYvQGQxwIx0CPBC3UKLR5gxh2eyWqWBYxmMOiJYyAA)

Inference from optional properties is a little bit weird in this sense. I'll have to think more about potential solutions for this on our side.

**Reply by AnatoleLucet** [reply]
Great! Seems to work fine when using `createActor`, but I get a new error when using it with `spawnChild` inside of a machine's `entry: ...`.

[Repro](https://www.typescriptlang.org/play?exactOptionalPropertyTypes=true#code/PTAEAEFMA8EMGMAuB5ADoglgewHawDYAKATlqpMYgJ4AqV5AzgFyiLECukAsAFAYC2qLJVABvUA0iJ2qADSh4xSLESQAgkmHyGqWAHccAYQAWGfABNQAX1AAzUv1AAiaA0QrITgNygQrepDMoACsAHQAnKEAjLy88LhurIGIoAC8ElIyABSivKD+jCyiNrAMYnn5oBg4qOyITBWVoAA+5TxNHXZYWAD8LG7E1QDmXo0dVmP5rew45pC21ZDmo+3WsrxWAJShisqqALIIpjiQORXxOKrQ9aA5VTV1aWI2W2kAfLe5q-m23SzVtUQoV+WFAPR6zgARrBiE51qstvCtiteJJpKgzqsEIhhEFxKpEhMeFsdkoPId4MdTl98pBLsQqP1dAYTGZzFknATEHCxFVzCxOcknNZNhtRTw4mTVBoccQslzNitdh4ZcJ5cl5OIAXUWDM5gsTpZkZK9upNHKuZr7oCil0sALhS9FbwgA)

Is this a mistake from my end?
